### ISSUE TITLE
links in twisted.web docs are broken

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4951]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)| @habnabit reported|
|-|-|
|Trac ID|trac#4951|
|Type|defect|
|Created|2011-03-15 05:35:37Z|
|Branch|https://github.com/twisted/twisted/tree/web-links-4951|

Spidered the most recent docs; all of these links are broken in the documentation. Format is the document with broken links, followed by the broken links.

```
http://twistedmatrix.com/documents/current/web/howto/resource-templates.html
  http://twistedmatrix.com/documents/10.2.0/api/twisted.web.server.Resource.html

http://twistedmatrix.com/documents/current/web/howto/using-twistedweb.html
  http://twistedmatrix.com/documents/10.2.0/api/twisted.application.internet.TCPServer.html

http://twistedmatrix.com/documents/current/web/howto/web-overview.html
  http://twistedmatrix.com/documents/10.2.0/api/twisted.web.server.NOT_DONE_YET.html

http://twistedmatrix.com/documents/current/web/howto/xmlrpc.html
  http://twistedmatrix.com/documents/10.2.0/api/twisted.web.resource.Resources.html
  http://twistedmatrix.com/documents/10.2.0/api/twisted.web.xmlrpc.Binary.html
  http://twistedmatrix.com/documents/10.2.0/api/twisted.web.xmlrpc.Boolean.html
  http://twistedmatrix.com/documents/10.2.0/api/twisted.web.xmlrpc.DateTime.html

http://twistedmatrix.com/documents/10.2.0/api/twisted.web.client.html
  http://twistedmatrix.com/documents/10.2.0/api/twisted.web.client.WebClientContextFactory%200.html
```

Attachments:

* [twisted.web-links.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d49/d49318e6d93feca4178ad412bf3786f8cde1c5e2/44bffd09f1eab98d40800b3b800d1d87ddc04a56.patch) (4766 bytes) - added by thijs on 2011-06-22 18:48:59Z - against r32149

<details><summary>Searchable metadata</summary>

```
trac-id__4951 4951
type__defect defect
reporter__habnabit habnabit
priority__normal normal
milestone__ 
branch__branches_web_links_4951 branches/web-links-4951
branch_author__thijs thijs
status__closed closed
resolution__fixed fixed
component__web web
keywords__documentation documentation
time__1300167337000000 1300167337000000
changetime__1309537425000000 1309537425000000
version__None None
owner__thijs thijs
cc__jknight cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_2"></a>|@habnabit commented|
|-|-|

The last link in this list is related to this new pydoctor issue:
https://bugs.launchpad.net/pydoctor/+bug/735275

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_3"></a>|@habnabit commented|
|-|-|

(In [31117]) Branching to 'docfix-4951'

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_4"></a>|@lvh commented|
|-|-|

1.In `http://twistedmatrix.com/documents/current/web/howto/resource-templates.html`:
  1. `http://twistedmatrix.com/documents/10.2.0/api/twisted.web.server.Resource.html` is because `t.w.s.Resource` is now `t.w.resource.Resource`
 2. In `http://twistedmatrix.com/documents/current/web/howto/using-twistedweb.html`:
  1. The `TCPServer` link is broken because [its implementation does evil hackery](http://twistedmatrix.com/trac/browser/tags/releases/twisted-10.2.0/twisted/application/internet.py) to produce the classes, which confuses pydoctor into thinking they don't exist.
 3. In `http://twistedmatrix.com/documents/current/web/howto/web-overview.html`:
  1. The link to NOT_DONE_YET doesn't work because pydoctor doesn't generate that API doc at all. This can be fixed by adding `__all__` to `twisted.web.server`. This is very similar to what happened to `t.w.guard` in [#3904](https://github.com/twisted/twisted/issues/3904).
 4. In `http://twistedmatrix.com/documents/current/web/howto/xmlrpc.html`:
  1. The link to Resources is just... wrong. Should point to the singular, Resource.
  2. The links to Fault et al. should just be removed or pointed to the stdlib xmlrpclib docs. `01:30 < glyph> I'm not so sure that having aliases for Fault et. al. is actually a good idea, though. So let's leave that out of the docs :)`
 5. Last one is already handled in previous comment.

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_5"></a>|@tenth10th commented|
|-|-|

(In [31278]) Branching to 'broken-doc-link-4951'

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_6"></a>|@tenth10th commented|
|-|-|

(Whoops, opened a branch with the wrong ticket number! Please disregard my changes here.)

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_7"></a>|@tenth10th commented|
|-|-|

Though, please note that [#11338](https://github.com/twisted/twisted/issues/11338), the broken link(s) in http://twistedmatrix.com/documents/current/web/howto/resource-templates.html are being addressed in [#4968](https://github.com/twisted/twisted/issues/4968).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

It seems no work was done in the branch associated with this ticket so I attached a new patch.

Replying to [lvh](#note_4):
>  1. In `http://twistedmatrix.com/documents/current/web/howto/resource-templates.html`:
>   1. `http://twistedmatrix.com/documents/10.2.0/api/twisted.web.server.Resource.html` is because `t.w.s.Resource` is now `t.w.resource.Resource`

fixed with attached patch.

>  2. In `http://twistedmatrix.com/documents/current/web/howto/using-twistedweb.html`:
>   1. The `TCPServer` link is broken because [its implementation does evil hackery](http://twistedmatrix.com/trac/browser/tags/releases/twisted-10.2.0/twisted/application/internet.py) to produce the classes, which confuses pydoctor into thinking they don't exist.

Let's ignore this then?

>  3. In `http://twistedmatrix.com/documents/current/web/howto/web-overview.html`:
>   1. The link to NOT_DONE_YET doesn't work because pydoctor doesn't generate that API doc at all. This can be fixed by adding `__all__` to `twisted.web.server`. This is very similar to what happened to `t.w.guard` in [#3904](https://github.com/twisted/twisted/issues/3904).

[#4966](https://github.com/twisted/twisted/issues/4966) was opened for this.

>  4. In `http://twistedmatrix.com/documents/current/web/howto/xmlrpc.html`:
>   1. The link to Resources is just... wrong. Should point to the singular, Resource.

fixed with attached patch.

>   2. The links to Fault et al. should just be removed or pointed to the stdlib xmlrpclib docs. `01:30 < glyph> I'm not so sure that having aliases for Fault et. al. is actually a good idea, though. So let's leave that out of the docs :)`

I'll remove it in a new branch I'll create for this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

(In [32152]) Branching to 'web-links-4951'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

(In [32153]) Apply twisted.web-links.patch, refs [#4951](https://github.com/twisted/twisted/issues/4951)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

(In [32185]) Remove xmlrpclib alias links, refs [#4951](https://github.com/twisted/twisted/issues/4951)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/builders/documentation/builds/1354). Up for review.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_15"></a>|@jesstess set owner to @thijstriemstra|
|-|-|

Thanks for getting this finished and the bonus doc standards cleanup, thijs.

 * The new http://wsgi.org link is a meta refresh to http://wsgi.org/wsgi/, which was the old link.
 * NOT_DONE_YET got a new ticket, but the TCPServer and WebClientContextFactory links from the ticket description are still broken. Either this ticket should stay open until they are fixed, or they should get their own tickets.

Other than that, looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_16"></a>|@thijstriemstra commented|
|-|-|

(In [32190]) address review comments, add news file. refs [#4951](https://github.com/twisted/twisted/issues/4951)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Replying to [jesstess](#note_15):
> Thanks for getting this finished and the bonus doc standards cleanup, thijs.
> 
>  * The new http://wsgi.org link is a meta refresh to http://wsgi.org/wsgi/, which was the old link.

Reverted this to the old link.

>  * NOT_DONE_YET got a new ticket, but the TCPServer and WebClientContextFactory links from the ticket description are still broken. Either this ticket should stay open until they are fixed, or they should get their own tickets.

I've removed the API link for `TCPServer` and I'll open a new ticket  for the `WebClientContextFactory` issue because it's probably not a strict doc issue and needs some code changes.

> 
> Other than that, looks great, please merge.

thanks for the review.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [32192]) Merge web-links-4951: Correct some broken links in the web howto documents.

Author: thijs
Reviewer: jesstess
Fixes: [#4951](https://github.com/twisted/twisted/issues/4951)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_19"></a>|@thijstriemstra commented|
|-|-|

Replying to [thijs](#note_17):
> I've removed the API link for `TCPServer`

Related ticket for this: [#4965](https://github.com/twisted/twisted/issues/4965)

> and I'll open a new ticket  for the `WebClientContextFactory` issue because it's probably not a strict doc issue and needs some code changes.

I'm not sure where to file this ticket, pydoctor or twisted..
